### PR TITLE
[Exchanges] support auto-filled exchanges

### PR DESCRIPTION
**Pull request by GuillaumeDSM**

### CODE REVIEW

File: octobot_trading/exchanges/util/exchange_util.py [hunk 1]
```diff
@@ -66,12 +70,66 @@ def get_exchange_class_from_name(exchange_parent_class, exchange_name, tentacles
...
+                aiohttp_session, auto_filled_exchanges[exchange_name][1], exchange_name
+            )
+    try:
+        exchange = getattr(ccxt, exchange_name)()
+        return exchange_details.ExchangeDetails(
+            exchange.id,
+            exchange.name,
+            exchange.urls["www"],
+            exchange.urls["api"],
+            exchange.urls["logo"],
```
Comment (Herklos) [inline octobot_trading/exchanges/util/exchange_util.py:121]: Shouldn't we create an enum for these strings?

File: tests/exchanges/util/test_exchange_util.py [hunk 2]
```diff
@@ -107,3 +117,126 @@ async def test_is_compatible_account_with_unchecked_exchange(exchange_config, te
...
+            get_tentacle_config_mock.assert_not_called()
+            assert exchanges.get_exchange_class_from_name(
+                exchanges.RestExchange, "blip", tentacles_setup_config, True,
+                strict_name_matching=True
+            ) == MockedAutoFillRestExchange
+            get_tentacle_config_mock.assert_called_once()
+
+
+@pytest.mark.asyncio
+async def test_get_exchange_details(tentacles_setup_config, supported_exchanges):
```
Comment (Herklos) [inline tests/exchanges/util/test_exchange_util.py:204]: :+1:

File: tests/exchanges/__init__.py [hunk 3]
```diff
@@ -41,6 +44,54 @@
...
+    @classmethod
+    def get_exchange_connector_class(cls, exchange_manager):
+        return cls.DEFAULT_CONNECTOR_CLASS
+
+    @classmethod
+    def is_default_exchange(cls) -> bool:
+        return False
+
+
+class MockedAutoFillRestExchange(MockedRestExchange):
```
Comment (Herklos) [inline tests/exchanges/__init__.py:65]: :+1:

**Review by Herklos — COMMENTED** [review]
Nice implementation!

File: octobot_trading/exchanges/util/exchange_util.py (hunk 1 above)
Comment (GuillaumeDSM) [reply octobot_trading/exchanges/util/exchange_util.py:121]: right

File: octobot_trading/exchanges/util/exchange_util.py [hunk 4]
```diff
@@ -66,12 +70,66 @@ def get_exchange_class_from_name(exchange_parent_class, exchange_name, tentacles
...
+        config = api.get_tentacle_config(
+            tentacles_setup_config, exchange_candidate
+        )
+        for exchange_name in exchange_candidate.supported_autofill_exchanges(config):
+            auto_filled_exchanges[exchange_name] = (exchange_candidate, config)
+    return auto_filled_exchanges
+
+
+def get_auto_filled_exchange_names(tentacles_setup_config):
+    return list(_get_auto_filled_exchanges(tentacles_setup_config))
```
[the remark this answers is not in the document]
Comment (Herklos) [reply octobot_trading/exchanges/util/exchange_util.py:95]: Oh yes my bad, I was looking at the wrong method

File: octobot_trading/exchanges/util/exchange_util.py [hunk 5]
```diff
@@ -66,12 +71,66 @@ def get_exchange_class_from_name(exchange_parent_class, exchange_name, tentacles
...
+        if exchange_name in auto_filled_exchanges:
+            return await auto_filled_exchanges[exchange_name][0].get_autofilled_exchange_details(
+                aiohttp_session, auto_filled_exchanges[exchange_name][1], exchange_name
+            )
+    try:
+        exchange = getattr(ccxt, exchange_name)()
+        return exchange_details.ExchangeDetails(
+            exchange.id,
+            exchange.name,
+            exchange.urls[ccxt_enums.ExchangeColumns.WEBSITE.value],
```
Comment (Herklos) [inline octobot_trading/exchanges/util/exchange_util.py:120]: :+1:

**Review by Herklos — APPROVED** [review]
:+1:

**Pull request merged**